pkg/controller/webpushvariant/webpushvariant_controller.go

```diff
@@ -0,0 +1,174 @@
+package webpushvariant
+
+import (
+	"context"
+	"time"
+
+	pushv1alpha1 "github.com/aerogear/unifiedpush-operator/pkg/apis/push/v1alpha1"
+	"github.com/aerogear/unifiedpush-operator/pkg/controller/util"
+	"github.com/aerogear/unifiedpush-operator/pkg/nspredicate"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+var log = logf.Log.WithName("controller_webpushvariant")
+
+// Add creates a new WebPushVariant Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileWebPushVariant{client: mgr.GetClient(), scheme: mgr.GetScheme()}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("webpushvariant-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource WebPushVariant
+	onlyEnqueueForValidNamespaces, err := nspredicate.NewFromEnvVar("APP_NAMESPACES")
+	if err != nil {
+		return err
+	}
+	err = c.Watch(
+		&source.Kind{Type: &pushv1alpha1.WebPushVariant{}},
+		&handler.EnqueueRequestForObject{},
+		onlyEnqueueForValidNamespaces,
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileWebPushVariant{}
+
+// ReconcileWebPushVariant reconciles a WebPushVariant object
+type ReconcileWebPushVariant struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a WebPushVariant
+// object and makes changes based on the state read and what is in the
+// WebPushVariant.Spec
+// Note:
+// The Controller will requeue the Request to be processed again if
+// the returned error is non-nil or Result.Requeue is true, otherwise
+// upon completion it will remove the work from the queue.
+func (r *ReconcileWebPushVariant) Reconcile(request reconcile.Request) (reconcile.Result, error) {
```

Comment:
In MSS we implemented some stuff to check the namespace that a consumption cr is found in. In this case we'd expect to action only crs created in MDC namespace. Do we need to do that here? Did we do it elsewhere in UPS? It might be moot because of permissions stuff changing but mentioning it just in case.

Reply:
Yes, we do that here, but the Reconcile is not really the best place for it: rather than calling this method for CRs in all namespaces and returning early if they're not in the approved set, we set a predicate on the watch so that it will only enqueue events that are from namespaces in that set. See lines 43 to 55 for where that's being done (and let me know if you want me to explain any of that :+1: ).

Reply:
Hey, that makes sense, I wonder if we can implement that in MSS, wdyt?

Reply:
Ok. I'll create a JIRA for it. It's probably low priority so I'll mark it as such. Will resolve this comment now.